File: robert-crypto-grpc-server/src/test/kotlin/fr/gouv/stopc/robertserver/crypto/ECDHKeyUsageDemonstrationTest.java
package fr.gouv.stopc.robertserver.crypto;

import org.apache.commons.lang3.builder.ToStringBuilder;
import org.apache.commons.lang3.builder.ToStringStyle;
import org.bouncycastle.pqc.math.linearalgebra.ByteUtils;
import org.junit.jupiter.api.Test;

import javax.crypto.KeyAgreement;

import java.security.KeyPairGenerator;
import java.security.spec.ECGenParameterSpec;

import static org.assertj.core.api.Assertions.*;

/**
 * Demonstrates ECDH key exchange.
 */
class ECDHKeyUsageDemonstrationTest {

    @Test
    void demonstrate_ecdh_key_derivation() throws Exception {

        final var kpg = KeyPairGenerator.getInstance("EC");
        kpg.initialize(new ECGenParameterSpec("secp256r1"));
        final var serverECDHKey = kpg.generateKeyPair();

        final var serverPrivateKey = serverECDHKey.getPrivate();
        assertThat(serverPrivateKey.getFormat()).isEqualTo("PKCS#8");
        assertThat(serverPrivateKey.getAlgorithm()).isEqualTo("EC");
        System.out.println("Server private key data as hex: " + ByteUtils.toHexString(serverPrivateKey.getEncoded()));
        System.out.println(
                "Server private key data as binary: " + ByteUtils.toBinaryString(serverPrivateKey.getEncoded())
        );
        System.out.println(ToStringBuilder.reflectionToString(serverPrivateKey, ToStringStyle.MULTI_LINE_STYLE));

        // use server public key as client public key
        final var clientPublicKey = serverECDHKey.getPublic();
        assertThat(clientPublicKey.getFormat()).isEqualTo("X.509");
        assertThat(clientPublicKey.getAlgorithm()).isEqualTo("EC");
        System.out.println("Server private key data as hex: " + ByteUtils.toHexString(clientPublicKey.getEncoded()));
        System.out.println(
                "Server private key data as binary: " + ByteUtils.toBinaryString(clientPublicKey.getEncoded())
        );
        System.out.println(ToStringBuilder.reflectionToString(clientPublicKey, ToStringStyle.MULTI_LINE_STYLE));

        final var ka = KeyAgreement.getInstance("ECDH");
        ka.init(serverPrivateKey);
        ka.doPhase(clientPublicKey, true);

        final var generatedSecret = ka.generateSecret();
        System.out.println(ToStringBuilder.reflectionToString(ka, ToStringStyle.MULTI_LINE_STYLE));
        System.out.println("Generated shared secret: " + ByteUtils.toHexString(generatedSecret));

    }
}

File: robert-crypto-grpc-server/src/test/kotlin/fr/gouv/stopc/robertserver/crypto/ECDHKeyUsageDemonstrationTest.kt
package fr.gouv.stopc.robertserver.crypto

import org.apache.commons.lang3.builder.ToStringBuilder
import org.apache.commons.lang3.builder.ToStringStyle
import org.assertj.core.api.Assertions.assertThat
import org.bouncycastle.pqc.math.linearalgebra.ByteUtils
import org.junit.jupiter.api.Test
import java.security.KeyPairGenerator
import java.security.spec.ECGenParameterSpec
import javax.crypto.KeyAgreement

/**
 * Demonstrates ECDH key exchange.
 */
internal class ECDHKeyUsageDemonstrationTest {

    @Test
    fun demonstrate_ecdh_key_derivation() {
        val kpg = KeyPairGenerator.getInstance("EC")
        kpg.initialize(ECGenParameterSpec("secp256r1"))
        val serverECDHKey = kpg.generateKeyPair()
        val serverPrivateKey = serverECDHKey.private
        assertThat(serverPrivateKey.format).isEqualTo("PKCS#8")
        assertThat(serverPrivateKey.algorithm).isEqualTo("EC")
        println("Server private key data as hex: " + ByteUtils.toHexString(serverPrivateKey.encoded))
        println("Server private key data as binary: " + ByteUtils.toBinaryString(serverPrivateKey.encoded))
        println(ToStringBuilder.reflectionToString(serverPrivateKey, ToStringStyle.MULTI_LINE_STYLE))

        // use server public key as client public key
        val clientPublicKey = serverECDHKey.public
        assertThat(clientPublicKey.format).isEqualTo("X.509")
        assertThat(clientPublicKey.algorithm).isEqualTo("EC")
        println("Server private key data as hex: " + ByteUtils.toHexString(clientPublicKey.encoded))
        println("Server private key data as binary: " + ByteUtils.toBinaryString(clientPublicKey.encoded))
        println(ToStringBuilder.reflectionToString(clientPublicKey, ToStringStyle.MULTI_LINE_STYLE))
        val ka = KeyAgreement.getInstance("ECDH")
        ka.init(serverPrivateKey)
        ka.doPhase(clientPublicKey, true)
        val generatedSecret = ka.generateSecret()
        println(ToStringBuilder.reflectionToString(ka, ToStringStyle.MULTI_LINE_STYLE))
        println("Generated shared secret: " + ByteUtils.toHexString(generatedSecret))
    }
}
